### DISCUSSION TITLE
I2C Device Driver 63 (HM330X) Missing?

### DISCUSSION THREAD

**Discussion by Teleportist** [arendst/Tasmota#18844] [Q&A]
### PROBLEM DESCRIPTION
HM330X connected to an ESP-32 via I2C has stopped producing telemetry.
I2CScan finds the device on bus1 0x40.
I2CDriver shows that the required driver 63 is missing? Not shown in result?
`RESULT = {"I2CDriver":"!7,8,!9,10,11,!12,13,!14,15,17,18,20,24,29,31,36,41,42,44,46,48,58,62,!65,69,76,77"}`

Have tried updating from 12.5.0 to 12.5.0.4. No fix.

Thanks

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [ X] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [X ] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X ] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X ] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [X ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X ] Device used (e.g., Sonoff Basic):EZP-32 + HM330X and SGP30
- [X ] Tasmota binary firmware version number used: 12.5.0 & 12.5.0.4
  - [X ] Pre-compiled
  - [ ] Self-compiled
- [X ] Flashing tools used: None. File upload to device
- [X ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
`13:03:17.404 CMD: Backlog Template; Module; GPIO 255
13:03:17.427 MQT: stat/Airmon/RESULT = {"NAME":"HM3301","GPIO":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,8160,0,640,608,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":1}
13:03:17.659 MQT: stat/Airmon/RESULT = {"Module":{"1":"ESP32-DevKit"}}
13:03:17.870 MQT: stat/Airmon/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"1216":"AM2301"},"GPIO5":{"0":"None"},"GPIO6":{"0":"None"},"GPIO7":{"0":"None"},"GPIO8":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO11":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"},"GPIO18":{"0":"None"},"GPIO19":{"8160":"BL0942 Rx"},"GPIO20":{"0":"None"},"GPIO21":{"640":"I2C SDA1"},"GPIO22":{"608":"I2C SCL1"},"GPIO23":{"0":"None"},"GPIO24":{"0":"None"},"GPIO25":{"0":"None"},"GPIO26":{"0":"None"},"GPIO27":{"0":"None"},"GPIO32":{"0":"None"},"GPIO33":{"0":"None"},"GPIO34":{"0":"None"},"GPIO35":{"0":"None"},"GPIO36":{"0":"None"},"GPIO37":{"0":"None"},"GPIO38":{"0":"None"},"GPIO39":{"0":"None"}}`
  Configuration output here:

```
- [ X] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here: None


```
- [ X] Provide the output of this command: `Status 0`:
```lua
`13:07:36.998 MQT: stat/Airmon/STATUS = {"Status":{"Module":1,"DeviceName":"Airmon","FriendlyName":["Airmon"],"Topic":"Airmon","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
13:07:37.006 MQT: stat/Airmon/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"https://github.com/arendst/Tasmota/files/7228154/tasmota-hm330x.zip","RestartReason":"Software reset CPU","Uptime":"0T00:20:57","StartupUTC":"2023-06-10T11:46:40","Sleep":50,"CfgHolder":4617,"BootCount":149,"BCResetTime":"2021-09-26T18:53:07","SaveCount":471}}
13:07:37.012 MQT: stat/Airmon/STATUS2 = {"StatusFWR":{"Version":"12.5.0.4(tasmota32)","BuildDateTime":"2023-06-10T09:32:45","Core":"2_0_9","SDK":"4.4.4.20230310","CpuFrequency":160,"Hardware":"ESP32-D0WDQ6 rev.1","CR":"391/699"}}
13:07:37.016 MQT: stat/Airmon/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A002800000000","00000080","00006000","00004000","00000000"]}}
13:07:37.032 MQT: stat/Airmon/STATUS4 = {"StatusMEM":{"ProgramSize":1526,"Free":1856,"Heap":111,"StackLowMark":3,"PsrMax":0,"PsrFree":0,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"1640D8","FlashFrequency":40,"FlashMode":"DIO","Features":["00000809","9F9AD7CF","0015A001","B7F7BFCF","05DA9BC0","E0360DC7","480840D2","20200000","D4BC482D","000A80B1"],"Drivers":"1,2,3,4,5,7,8,9,10,11,12,14,16,17,20,21,24,26,27,29,34,35,38,50,52,59,60,62,63,66,67,68,82,86,87,88","Sensors":"1,2,3,5,6,7,8,9,10,11,12,13,14,15,17,18,19,20,21,22,26,31,34,37,39,40,42,43,45,51,52,55,56,58,59,64,66,67,74,85,92,95,98,103,105,127","I2CDriver":"!7,8,!9,10,11,!12,13,!14,15,17,18,20,24,29,31,36,41,42,44,46,48,58,62,!65,69,76,77"}}
13:07:37.044 MQT: stat/Airmon/STATUS5 = {"StatusNET":{"Hostname":"Airmon","IPAddress":"192.168.1.234","Gateway":"192.168.1.1","Subnetmask":"255.255.255.0","DNSServer1":"192.168.1.1","DNSServer2":"0.0.0.0","Mac":"","IP6Global":"","IP6Local":"","Ethernet":{"Hostname":"","IPAddress":"0.0.0.0","Gateway":"0.0.0.0","Subnetmask":"0.0.0.0","DNSServer1":"192.168.1.1","DNSServer2":"0.0.0.0","Mac":"00:00:00:00:00:00","IP6Global":"","IP6Local":""},"Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
13:07:37.051 MQT: stat/Airmon/STATUS6 = {"StatusMQT":{"MqttHost":"192.168.1.137","MqttPort":1883,"MqttClientMask":"Airmon","MqttClient":"Airmon","MqttUser":"hassio","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
13:07:37.058 MQT: stat/Airmon/STATUS7 = {"StatusTIM":{"UTC":"2023-06-10T12:07:37","Local":"2023-06-10T13:07:37","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":"+01:00","Sunrise":"04:47","Sunset":"20:51"}}
13:07:37.069 MQT: stat/Airmon/STATUS10 = {"StatusSNS":{"Time":"2023-06-10T13:07:37","AM2301":{"Temperature":26.5,"Humidity":46.1,"DewPoint":14.0},"SGP30":{"eCO2":403,"TVOC":68,"aHumidity":10.9800},"TempUnit":"C"}}
13:07:37.079 MQT: stat/Airmon/STATUS11 = {"StatusSTS":{"Time":"2023-06-10T13:07:37","Uptime":"0T00:20:57","UptimeSec":1257,"Heap":113,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":20,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":36},"Wifi":{"AP":1,"SSId":"","BSSId":"","Channel":6,"Mode":"11n","RSSI":92,"Signal":-54,"LinkCount":1,"Downtime":"0T00:00:05"}}}`

**Comment by barbudor** [comment]
Drivers do not disappear magically
You upgraded from a self compiled binary with HM330x driver to a standard build without that driver installed
Hm33x is NOT included in any standard build, you have to compile it yourself

**Reply by Teleportist** [reply]
Thank you Barbudor,

You'll of course remember kindly compiling an older version of Tasmota with this driver for me, which worked well and was most appreciated.
I had (wrongly) assumed that this driver would be rolled into future releases of Tasmota. I'm assuming drivers are left out to save space?

So now I will attempt do downgrade my version, back to the one you provided.
I might be tempted to try and compile an up to date version. Are you aware of any written of videos, with step-by-step instructions for dummies, that could help me achieve this please?

**Comment by Teleportist** [comment]
So I seem to have resolved this.

I began by trying to compile a version, as per this guide> [https://www.youtube.com/watch?v=WashxTcHiDc](url)
Unfortunately, flashing the file didn't give me the functionality I wanted. Maybe I didn't make the correct changes?

![image](https://github.com/arendst/Tasmota/assets/61894621/55cf42d2-0f0d-42e7-a419-ca48b8af1325)

Then I just happened to notice the option in the Tasmota web flasher for a (unofficial) binary with all sensors. Installing this worked perfectly.
It's probably a lot more bloated than need be, but suits my needs for now.

![image](https://github.com/arendst/Tasmota/assets/61894621/4eea41d4-b3dd-415c-9c80-d49238ca2dcd)
